src/plugins/plugin.legend.js

```diff
@@ -459,58 +465,75 @@ var Legend = Element.extend({
 		}
 	},
 
+	/**
+	 * Gets the legend item, if any, at a given x,y position.
+	 * @private
+	 * @param {number} x - The x coordinate to check
+	 * @param {number} y - The y coordindate to check
+	 * @return {?Object} - The legend item found
+	 */
+	getLegendItemAt: function(x, y) {
+		var me = this;
+
+		if (x >= me.left && x <= me.right && y >= me.top && y <= me.bottom) {
+			// See if we are touching one of the dataset boxes
+			var lh = me.legendHitBoxes;
+			for (var i = 0; i < lh.length; ++i) {
+				var hitBox = lh[i];
+
+				if (x >= hitBox.left && x <= hitBox.left + hitBox.width && y >= hitBox.top && y <= hitBox.top + hitBox.height) {
+					// Touching an element
+					return me.legendItems[i];
+				}
+			}
+		}
+
+		return null;
+	},
+
 	/**
 	 * Handle an event
 	 * @private
 	 * @param {IEvent} event - The event to handle
-	 * @return {Boolean} true if a change occured
 	 */
 	handleEvent: function(e) {
 		var me = this;
 		var opts = me.options;
 		var type = e.type === 'mouseup' ? 'click' : e.type;
-		var changed = false;
 
 		if (type === 'mousemove') {
-			if (!opts.onHover) {
+			if (!opts.onHover && !opts.onLeave) {
 				return;
 			}
 		} else if (type === 'click') {
 			if (!opts.onClick) {
 				return;
 			}
-		} else {
+		}	else {
 			return;
 		}
 
 		// Chart event already has relative position in it
-		var x = e.x;
-		var y = e.y;
-
-		if (x >= me.left && x <= me.right && y >= me.top && y <= me.bottom) {
-			// See if we are touching one of the dataset boxes
-			var lh = me.legendHitBoxes;
-			for (var i = 0; i < lh.length; ++i) {
-				var hitBox = lh[i];
+		var hoveredItem = me.getLegendItemAt(e.x, e.y);
 
-				if (x >= hitBox.left && x <= hitBox.left + hitBox.width && y >= hitBox.top && y <= hitBox.top + hitBox.height) {
-					// Touching an element
-					if (type === 'click') {
-						// use e.native for backwards compatibility
-						opts.onClick.call(me, e.native, me.legendItems[i]);
-						changed = true;
-						break;
-					} else if (type === 'mousemove') {
-						// use e.native for backwards compatibility
-						opts.onHover.call(me, e.native, me.legendItems[i]);
-						changed = true;
-						break;
-					}
+		if (type === 'click') {
+			if (hoveredItem && opts.onClick) {
+				// use e.native for backwards compatibility
+				opts.onClick.call(me, e.native, hoveredItem);
+			}
+		} else {
```

Comment:
I wonder if you still need to check for `type === 'mousemove'` here?  E.g. could this method be called for touch events? https://www.chartjs.org/docs/latest/general/interactions/events.html

Reply:
Wouldn't such events be excluded by the logic at the top of the method? E.g. any event with a type other than `mousemove` and `click` should have already returned, as I understand it?

Reply:
> Wouldn't such events be excluded by the logic at the top of the method?

No, it's not, this method [receives all DOM events](/chartjs/Chart.js/blob/master/src/plugins/plugin.legend.js#L570) so it still needs to be filtered.

Reply:
I know, but this the logic at the top of the method is:

```ts
		if (type === 'mousemove') {
			if (!opts.onHover && !opts.onLeave) {
				return;
			}
		} else if (type === 'click') {
			if (!opts.onClick) {
				return;
			}
		}	else {
			return;
		}
```

How can a type other than `mousemove` or `click` would get through this without returning?

Reply:
@jonrimmer you right (I missed the part of the code you were referring to)

Not necessary for this PR but I would refactor the event logic in separate methods to make clear the event scope and avoid check multiple time the event type:

```javascript
/**
 * @private
 */
_handleMoveEvent: function(e) {
    // ...
}
/**
 * @private
 */
_handleClickEvent: function(e) {
    // ...
}
handleEvent: function(e) {
    switch (e.type) {
    case 'mouseup':
        this._handleClickEvent(e);
        break;
    case 'mousemove'
        this._handleMoveEvent(e);
        break,
    default:
    }
}
```

Reply:
(I deleted my previous message before realizing you already replied to it)